src/control/system/checker/finding.go

```diff
@@ -0,0 +1,146 @@
+}
+
+func descAction(action chkpb.CheckInconsistAction, ro reportObject, details ...string) string {
+	// Create a map of details. If a detail is not found by
+	// the expected index, then the default is an empty string.
+	detMap := make(map[int]string)
+	for i, det := range details {
+		detMap[i] = det
+	}
+
+	switch action {
+	case chkpb.CheckInconsistAction_CIA_IGNORE:
+		return fmt.Sprintf("Ignore the %s finding", ro)
+	case chkpb.CheckInconsistAction_CIA_DISCARD:
+		return fmt.Sprintf("Discard the %q %s", detMap[0], ro)
+	case chkpb.CheckInconsistAction_CIA_READD:
+		return fmt.Sprintf("Re-add the %q %s", detMap[0], ro)
+	case chkpb.CheckInconsistAction_CIA_TRUST_MS:
+		return fmt.Sprintf("Trust the MS %s entry (%s) for %s", ro, detMap[1], detMap[0])
+	case chkpb.CheckInconsistAction_CIA_TRUST_PS:
+		return fmt.Sprintf("Trust the PS %s entry (%s) for %s", ro, detMap[1], detMap[0])
+	case chkpb.CheckInconsistAction_CIA_TRUST_TARGET:
+		return fmt.Sprintf("Trust the %s result (%s) for %s", ro, detMap[1], detMap[0])
+	case chkpb.CheckInconsistAction_CIA_TRUST_MAJORITY:
+		return fmt.Sprintf("Trust the majority of the %s results (%s) for %s", ro, detMap[1], detMap[0])
+	case chkpb.CheckInconsistAction_CIA_TRUST_LATEST:
+		return fmt.Sprintf("Trust the latest %s result (%s) for %s", ro, detMap[1], detMap[0])
+	case chkpb.CheckInconsistAction_CIA_TRUST_OLDEST:
+		return fmt.Sprintf("Trust the oldest %s result (%s) for %s", ro, detMap[1], detMap[0])
+	case chkpb.CheckInconsistAction_CIA_TRUST_EC_PARITY:
+		return fmt.Sprintf("Trust the parity of the %s results (%s) for %s", ro, detMap[1], detMap[0])
+	case chkpb.CheckInconsistAction_CIA_TRUST_EC_DATA:
+		return fmt.Sprintf("Trust the data of the %s results (%s) for %s", ro, detMap[1], detMap[0])
+	default:
+		return fmt.Sprintf("%s: %s (details: %+v)", ro, action, details)
+	}
+}
+
+func AnnotateFinding(f *Finding) *Finding {
+	if f == nil {
+		return nil
+	}
+
+	// Pad out the list of details as necessary to match
+	// the length of the action list.
+	if len(f.Details) != len(f.Actions) {
+		for i := len(f.Details); i < len(f.Actions); i++ {
+			f.Details = append(f.Details, "")
+		}
+	}
+
+	switch f.Class {
+	case chkpb.CheckInconsistClass_CIC_POOL_NONEXIST_ON_MS:
+		if f.Msg == "" {
+			f.Msg = fmt.Sprintf("Scanned pool service %s missing from MS", f.PoolUuid)
+		}
+		for i, act := range f.Actions {
+			f.Details[i] = descAction(act, poolObj, f.PoolUuid)
+		}
+	case chkpb.CheckInconsistClass_CIC_POOL_NONEXIST_ON_ENGINE:
+		if f.Msg == "" {
+			f.Msg = fmt.Sprintf("MS pool service %s missing on engines", f.PoolUuid)
+		}
+		for i, act := range f.Actions {
+			f.Details[i] = descAction(act, poolObj, f.PoolUuid)
+		}
+	case chkpb.CheckInconsistClass_CIC_POOL_BAD_LABEL:
+		if f.Msg == "" {
+			f.Msg = fmt.Sprintf("The pool label for %s does not match MS", f.PoolUuid)
+		}
+		for i, act := range f.Actions {
+			f.Details[i] = descAction(act, poolObj, f.PoolUuid, f.Details[i])
```

Comment:
If I'm understanding what's happening here, we're overwriting the contents of the Details array with our formatted action string. IMO these should be a separate field, or maybe have different structs (internal vs. external, with the external having the full rendering of the string). Overloading seems destined to confuse.

Reply:
... Yeah, I think you're right. The details array is supplied by the originator of the report (e.g. the in-engine checker module) and the idea is that we wanted to handle formatting of human-readable messages in the control plane instead of the engine. I agree that it would probably be best to preserve the original details array and add a new array of full message strings.

Reply:
My instinct would be to push the string rendering up to the control API layer--maybe that's a motivation to define a separate non-protobuf data structure at the control API level.